### PR TITLE
{cae} [intel-2015b] OpenFOAM 3.0.0. external dependencies (REVIEW)

### PR DESCRIPTION
**Pull request by hajgato**
Needs PR ~~#2136~~ & ~~#2139~~

needs ~~https://github.com/hpcugent/easybuild-easyblocks/pull/744~~

### CODE REVIEW

**Comment by hajgato** [comment]
64 bit integer version (`toolchainopts += {'í8': True}`) does not work with `METIS`. `SCOTCH` works with the 64bitint bit version.

**Comment by boegel** [comment]
Jenkins: test this please

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps.eb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    ('cleanup-ThirdParty-%(version)s.patch', ".."),  # patch should not be applied in OpenFOAM subdir
+    'external-3rd-OpenFOAM-%(version)s.patch',
+    ('external-3rd-ThirdParty-%(version)s.patch', ".."),
+    'OpenFOAM-%s_external_paraview-%s.patch' % (version, paraversion),
+]
+
+dependencies = [
+    ('CGAL', '4.7', '-Python-2.7.10'),
+    ('libreadline', '6.3'),
+# OpenFOAM requires 64 bit METIS using 32 bit indexes (array indexes)
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps.eb]: fix indent please

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps.eb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    'ThirdParty-%(version)s.tgz',
+]
+
+patches = [
+    'cleanup-OpenFOAM-%(version)s.patch',
+    'OpenFOAM-3.0.0_libreadline.patch',
+    ('cleanup-ThirdParty-%(version)s.patch', ".."),  # patch should not be applied in OpenFOAM subdir
+    'external-3rd-OpenFOAM-%(version)s.patch',
+    ('external-3rd-ThirdParty-%(version)s.patch', ".."),
+    'OpenFOAM-%s_external_paraview-%s.patch' % (version, paraversion),
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps.eb]: use `%%(version)s`

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps.eb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+name = 'OpenFOAM'
+version = '3.0.0'
+versionsuffix = '-eb-deps'
+paraversion = '4.4.0'
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps.eb]: drop this down to above `patches` (first location where it's needed)

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ 
+ # set paths if binaries or source are present
+ if [ -r $ParaView_DIR -o -r $paraviewInstDir ]
+--- OpenFOAM-3.0.0/applications/utilities/postProcessing/graphics/PV4Readers/Allwmake.orig	2015-11-03 11:30:23.000000000 +0100
++++ OpenFOAM-3.0.0/applications/utilities/postProcessing/graphics/PV4Readers/Allwmake	2015-11-16 11:47:06.130326063 +0100
+@@ -12,7 +12,6 @@
+         }
+ 
+         # ensure CMake gets the correct C++ compiler
+-        [ -n "$WM_CXX" ] && export CXX="$WM_CXX"
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch]: is this a warning flag what we should be setting `$WM_CXX` in the OpenFOAM easyblock?

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_libreadline.patch [hunk 5]
```diff
@@ -0,0 +1,16 @@
+--- OpenFOAM-3.0.0/applications/utilities/mesh/manipulation/setSet/Allwmake.orig	2015-11-03 11:30:23.000000000 +0100
++++ OpenFOAM-3.0.0/applications/utilities/mesh/manipulation/setSet/Allwmake	2015-11-16 11:13:49.318760933 +0100
+@@ -6,11 +6,11 @@
+ #
+ # use readline if available
+ #
+-if [ -f /usr/include/readline/readline.h ]
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_libreadline.patch]: omfg, these people... >_<

File: easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+--- OpenFOAM-3.0.0/etc/config/CGAL.csh	2015-11-16 13:54:59.897564794 +0100
++++ OpenFOAM-3.0.0.orig/etc/config/CGAL.csh	2015-11-16 13:46:45.131988362 +0100
+@@ -29,26 +29,36 @@
+ #     Sourced from OpenFOAM-<VERSION>/etc/cshrc
+ ##------------------------------------------------------------------------------
+ 
+-set boost_version=boost-system
+-set cgal_version=CGAL-4.7
++if ( ${?EBROOTCGAL} != 0 ) then
++   setenv BOOST_ARCH_PATH $EBROOTBOOST
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch]: should this be guarded with a check for `$EBROOTBOOST`?

File: easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
++   setenv CGAL_ARCH_PATH $EBROOTCGAL
++
++elif ( ${?CGAL_ROOT} != 0 ) then
++   setenv BOOST_ARCH_PATH $CGAL_ROOT
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch]: this should use `$BOOST_ROOT`?

File: easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
++++ OpenFOAM-3.0.0.orig/etc/config/CGAL.sh	2015-11-03 11:30:23.000000000 +0100
+@@ -29,29 +29,39 @@
+ #     Sourced from OpenFOAM-<VERSION>/etc/bashrc
+ #------------------------------------------------------------------------------
+
+-boost_version=boost-system
+-cgal_version=CGAL-4.7
++if [ -n "${EBROOTCGAL}" ]; then
++    export CGAL_ARCH_PATH="${EBROOTCGAL}"
++    export BOOST_ARCH_PATH="${EBROOTBOOST}"
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch]: guard for `$EBROOTBOOST`?

File: easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
++
++elif [ -n "${CGAL_ROOT}" ]; then
++    export CGAL_ARCH_PATH="${CGAL_ROOT}"
++    export BOOST_ARCH_PATH="${CGAL_ROOT}"
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/external-3rd-OpenFOAM-3.0.0.patch]: `$BOOST_ROOT`?

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
Linux centos linux 7.1.1503, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/d31c60222319e55f688e for a full test report.

**Comment by boegel** [comment]
@hajgato: can you fix the minor remarks here, and in the easyblocks PR? works on my end, and we have a need for this on Tier1

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch (hunk 4 above)
Comment (hajgato) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch]: @boegel No. We CXX is set by EB, there is no need to set here. We do not use custom C++ compiler, so this is why I deleted this line.

**Comment by hajgato** [comment]
@boegel: I implemeted the requested changes. I also added the `Python-2.7.10` suffix (as for `OpenFOAM-2.{3.1,4.0}-eb-deps-Python-2.7.10`, but I am not sure that this was really needed.

**Comment by boegel** [comment]
Jenkins: test this please

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch (hunk 4 above)
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch]: OK, another option would have been to (also) define `$WM_CXX` in the OpenFOAM easyblock, but fine.

**Comment by boegel** [comment]
@hajgato:

```
AssertionError: Filename '.../OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps-Python-2.7.10.eb' 
of parsed easyconfig matches expected filename 'OpenFOAM-3.0.0-intel-2015b-eb-deps-Python-2.7.10).eb'
```

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps-Python-2.7.10.eb [hunk 10]
```diff
@@ -0,0 +1,47 @@
+name = 'OpenFOAM'
+version = '3.0.0'
+pyver = '-Python-2.7.10'
+versionsuffix = '-eb-deps%s)' % pyver
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps-Python-2.7.10.eb]: you probably don't want the `)` here

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch (hunk 4 above)
Comment (hajgato) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0_external_paraview-4.4.0.patch]: @boegel The easyblock `WM_CXX` would be overwritten by `settings.{sh,csh}` so then we would also have to patch that part of `settings`.

File: easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps-Python-2.7.10.eb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+source_urls = ['http://downloads.sourceforge.net/foam/%(version)s']
+sources = [
+    SOURCE_TGZ,
+    'ThirdParty-%(version)s.tgz',
+]
+
+paraversion = '4.4.0'
+patches = [
+    'OpenFOAM-%(version)s_cleanup.patch',
+    'OpenFOAM-%(version)_libreadline.patch',
```
Comment (boegel) [inline easybuild/easyconfigs/o/OpenFOAM/OpenFOAM-3.0.0-intel-2015b-eb-deps-Python-2.7.10.eb]: `%(version)s` (missing `s` at the end)

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (2 easyconfigs in this PR)
Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/a3c56f3c4f23b7e2f8bc for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (2 easyconfigs in this PR)
Linux centos linux 7.1.1503, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/b6579a37c0317e16a4f9 for a full test report.

**Comment by boegel** [comment]
Going in (since we need this ;-)), thanks @hajgato!

**Pull request merged**